lib/Sema/CSApply.cpp

```diff
@@ -7043,6 +7044,38 @@ Expr *ExprRewriter::coerceToType(Expr *expr, Type toType,
       }
     }
 
+    /// Whether the given effect should be propagated to a closure expression.
+    auto shouldApplyEffect = [&](EffectKind kind) -> bool {
+      auto last = locator.last();
+      if (!(last && last->getKind() == ConstraintLocator::ApplyArgToParam))
+        return true;
+
+      // The effect should not be applied if the closure is an argument
+      // to a function where that effect is polymorphic.
+      if (auto *call = getAsExpr<ApplyExpr>(locator.getAnchor())) {
+        if (auto *declRef = dyn_cast<DeclRefExpr>(call->getFn())) {
```

Comment:
Would a more robust way of determining this been getting a callee locator for the argument and getting a declaration via `getOverloadChoice`? That would handle situations where call was to an overloaded reference which is a possibility right?

Reply:
I tried using `getOverloadChoice` and it didn't work because the anchor is fully type-checked here. Given that it's fully type-checked, I think that should already account for overloads

Reply:
That's fortunate! Yeah, that should already have overload applied, not all of the places in CSApply behave that way so it's hard to say what is what :)